app/models/mentor_management.rb

```diff
@@ -0,0 +1,122 @@
+class MentorManagement
+  # Select a mentor using the following algorithm
+  #
+  # 1) Find all assignment participants for the
+  #    assignment with id [assignment_id] whose
+  #    duty is the same as [Particpant#DUTY_MENTOR].
+  # 2) Count the number of teams those participants
+  #    are a part of, acting as a proxy for the
+  #    number of teams they mentor.
+  # 3) Return the mentor with the fewest number of
+  #    teams they're currently mentoring.
+  #
+  # This method's runtime is O(n lg n) due to the call to
+  # Hash#sort_by. This assertion assumes that the
+  # database management system is capable of fetching the
+  # required rows at least as quickly.
+  #
+  # Implementation detail: Any tie between the top 2
+  # mentors is decided by the Hash#sort_by algorithm.
+  #
+  # @return The id of the mentor with the fewest teams
+  #   they are assigned to. Returns `nil` if there are
+  #   no participants with mentor duty for [assignment_id].
+  def self.select_mentor(assignment_id)
+    mentor_user_id, _ = self.zip_mentors_with_team_count(assignment_id).first
+    User.where(id: mentor_user_id).first
+  end
+
+  # = Mentor Management
+  # E2115: Handles calls when an assignment has the auto_assign_mentor flag enabled and triggered by the event when a new member joins an assignment team.
+  #
+  # This event happens when:
+  #   1.) An invited student user accepts and successfully added to a team from
+  #       app/models/invitation.rb
+  #   2.) A student user is successfully added to the team manually from
+  #       app/controllers/teams_users_controller.rb.
+  #
+  # This method will determine if a mentor needs to be assigned, if so,
+  # selects one, and adds the mentor to the team if:
+  #   1.) The assignment does not have a topic.
+  #   2.) If the team has reached >50% full capacity.
+  #   3.) If the team does not have a mentor.
+  def self.assign_mentor(assignment_id, team_id)
+    assignment = Assignment.find(assignment_id)
+    team = Team.find(team_id)
+
+    # RuboCop 'use guard clause instead of nested conditionals'
+    # return if assignments can't accept mentors
+    return unless assignment.auto_assign_mentor
+
+    # RuboCop 'use guard clause instead of nested conditionals'
+    # return if the assignment or team already have a topic
+    return if assignment.topics? || !team.topic.nil?
+
+    curr_team_size = Team.size(team_id)
+    max_team_members = Assignment.find(assignment_id).max_team_size
+
+    # RuboCop 'use guard clause instead of nested conditionals'
+    # return if the team size hasn't reached > 50% of capacity
+    return if curr_team_size * 2 <= max_team_members
+
+    # RuboCop 'use guard clause instead of nested conditionals'
+    # return if there's already a mentor in place
+    return if team.participants.any? { |participant| participant.duty == Participant::DUTY_MENTOR }
```

Comment:
why arent you using the def self.user_a_mentor?(user) function here?

Reply:
Fair point! `user_a_mentor?` checks whether there's a corresponding entry in `Participant` for a given `User` with the `mentor` duty. So we could equivalently do:
```ruby
return if team.users.any? &:user_a_mentor?
```

An advantage I can think of for the way it exists currently is that we avoid a bunch of database calls.

Reply:
i think this looks cleaner.